### ISSUE TITLE
open pragma with :std and :crlf

### ISSUE THREAD

**Issue by mik2** [Perl/perl5#18470]
Module: open

**Description**
```
open ":std", IO => ":crlf"
```
does not affect STDIN or STDOUT, which is in contradiction with the following statement in the documentation:

> :std is not a layer but an additional subpragma. When specified in the import list, it activates an additional functionality of pushing the layers selected for input/output handles to the standard filehandles (STDIN, STDOUT, STDERR)

**Steps to Reproduce**
```
printf "\r\n\n" | perl -Mopen=:std,IO,:crlf -e 'print <>;' | od -c
0000000  \r  \n  \n
0000003
```

**Expected behavior**
the output should be the same as for:
```
printf "\r\n\n" | perl -e 'binmode $_,":crlf" for STDIN, STDOUT; print <>;' | od -c
0000000  \r  \n  \r  \n
0000004
```

**Perl configuration**
```
Summary of my perl5 (revision 5 version 32 subversion 0) configuration:

  Platform:
    osname=linux
    osvers=5.4.43-1-lts
    archname=x86_64-linux-thread-multi
    uname='linux build-edge-x86_64 5.4.43-1-lts #2-alpine smp thu, 28 may 2020 20:13:48 utc x86_64 linux '
    config_args='-des -Dcccdlflags=-fPIC -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dprefix=/usr -Dprivlib=/usr/share/perl5/core_perl -Darchlib=/usr/lib/perl5/core_perl -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/vendor_perl -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5/site_perl -Dsitearch=/usr/local/lib/perl5/site_perl -Dlocincpth=  -Doptimize=-Os -fomit-frame-pointer -Duselargefiles -Dusethreads -Duseshrplib -Dd_semctl_semun -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dinstallman1dir=/usr/share/man/man1 -Dinstallman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3pm -Dcf_by=Alpine -Ud_csh -Dusenm'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-Os -fomit-frame-pointer'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong'
    ccversion=''
    gccversion='10.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/include/fortify /usr/lib /usr/local/lib /lib/../lib /usr/lib/../lib /lib
    libs=-lpthread -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/usr/lib/libc.a
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/core_perl/CORE'
    cccdlflags='-fPIC'
    lddlflags='-shared -Os -fomit-frame-pointer -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_IMPLICIT_CONTEXT
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at Sep 21 2020 17:38:49
  @INC:
    /usr/local/lib/perl5/site_perl
    /usr/local/share/perl5/site_perl
    /usr/lib/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/core_perl
    /usr/share/perl5/core_perl
```

**Comment by mik2** [comment]
According to documentation:

> :std is not a layer but an additional subpragma. When specified in the import list, it activates an additional functionality of pushing the layers selected for input/output handles to the standard filehandles (STDIN, STDOUT, STDERR)

**Comment by Leont** [comment]
Apparently, `open.pm`'s `:std` option is broken when it comes to layers without an argument. It will actually work if you do this: printf `"\r\n\n" | perl '-Mopen=:std,IO,:crlf(foo)' -e 'print <>;' | od -c` (though the argument is entirely nonsensical).

I have no idea why it works that way, the simpler thing of just binmoding the argument as-is would not have this issue.

**Comment by mik2** [comment]
> Apparently, `open.pm`'s `:std` option is broken when it comes to layers without an argument. It will actually work if you do this: printf `"\r\n\n" | perl '-Mopen=:std,IO,:crlf(foo)' -e 'print <>;' | od -c` (though the argument is entirely nonsensical).

thanks for a nice workaround

> I have no idea why it works that way, the simpler thing of just binmoding the argument as-is would not have this issue.

but this way you need to do both "use open" and "binmode" if you use "<>"

**Comment by Leont** [comment]
> but this way you need to do both "use open" and "binmode" if you use "<>"

I meant in the [open.pm code](https://github.com/Perl/perl5/blob/blead/lib/open.pm#L129), not in your code.

**Comment by Leont** [comment]
This was fixed in #18516
